spinalcordtoolbox/reports/qc.py

```diff
@@ -264,85 +281,127 @@ def wrapped_f(sct_slice, *args):
-            elif sct_slice.get_name() == 'Sagittal':
-                size_fig = [5 * img.shape[1] / img.shape[0], 5]
             fig.set_size_inches(size_fig[0], size_fig[1], forward=True)
             FigureCanvas(fig)
             ax = fig.add_axes((0, 0, 1, 1))
-            ax.imshow(img, cmap='gray', interpolation=self.interpolation, aspect=float(aspect_img))
-            self._add_orientation_label(ax)
-            ax.get_xaxis().set_visible(False)
-            ax.get_yaxis().set_visible(False)
-            self._save(fig, self.qc_report.qc_params.abs_bkg_img_path(), dpi=self.qc_report.qc_params.dpi)
-
-            for action in self.action_list:
-                logger.debug('Action List %s', action.__name__)
-                if self._stretch_contrast and action.__name__ in ("no_seg_seg",):
-                    print("Mask type %s" % mask.dtype)
-                    mask = func_stretch_contrast[self._stretch_contrast_method](mask)
-                fig = Figure()
-                fig.set_size_inches(size_fig[0], size_fig[1], forward=True)
-                FigureCanvas(fig)
-                ax = fig.add_axes((0, 0, 1, 1))
-                action(self, mask, ax)
-                self._save(fig, self.qc_report.qc_params.abs_overlay_img_path(), dpi=self.qc_report.qc_params.dpi)
-
-            self.qc_report.update_description_file(img.shape)
+            action(self, mask, ax)
+            self._save(fig, self.qc_report.qc_params.abs_overlay_img_path(), dpi=self.qc_report.qc_params.dpi)
 
-        return wrapped_f
+        self.qc_report.update_description_file(img.shape)
+
+    def _make_QC_image_for_4d_volumes(self, images_after_moco, images_before_moco):
+        """
+        Generate background and overlay gifs for sct_fmri_moco and sct_dmri_moco
+
+        :param images_after_moco: list of mosaic images after motion correction
+        :param images_before_moco: list of mosaic images before motion correction
+        :return:
+        """
+
+        size_fig = [5, 10 * images_after_moco[0].shape[0] / images_after_moco[0].shape[1] + 0.5]
+
+        for i in range(len(images_after_moco)):
+            images_after_moco[i] = self._func_stretch_contrast(images_after_moco[i])
+            images_before_moco[i] = self._func_stretch_contrast(images_before_moco[i])
+
+        self._generate_and_save_gif(images_before_moco, images_after_moco, size_fig)
+        self._generate_and_save_gif(images_before_moco, images_after_moco, size_fig, is_mask=True)
+
+        w, h = (self.qc_report.qc_params.dpi * size_fig[0], self.qc_report.qc_params.dpi * size_fig[1])
+        self.qc_report.update_description_file((w, h))
+
+    def _func_stretch_contrast(self, img):
+        if self._stretch_contrast_method == "equalized":
+            return self._equalize_histogram(img)
+        else:
+            p2, p98 = np.percentile(img, (2, 98))
+            return skimage.exposure.rescale_intensity(img, in_range=(p2, p98))
```

Comment:
I think this might behave differently than the existing code.

* Before, there were exactly 2 options: `equalized` and `contrast_stretching`. If a 3rd value was passed, then an error would occur.
* Now, if a 3rd value is passed, it will fall into the `else:` case, which might be unexpected.

Because of this, I think it would be good to explicitly check the value of `stretch_contrast_method` to ensure that `equalized` and `contrast_stretching` are the only valid options. Maybe in `__init__`, so that it's caught right away?

Reply:
Also, one small nitpicky thing: it could be good to also split the two `contrast_stretching` lines into a separate method like you did with `self._equalize_histogram`.